src/FSharpLint.Core/Rules/Conventions/EnsureTailCallDiagnosticsInRecursiveFunctions.fs

```diff
@@ -0,0 +1,65 @@
+﻿module FSharpLint.Rules.EnsureTailCallDiagnosticsInRecursiveFunctions
+
+open System
+
+open FSharpLint.Framework
+open FSharpLint.Framework.Suggestion
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Text
+open FSharp.Compiler.CodeAnalysis
+open FSharpLint.Framework.Ast
+open FSharpLint.Framework.Rules
+
+let runner (args: AstNodeRuleParams) =
+    match args.AstNode, args.CheckInfo with
+    | AstNode.ModuleDeclaration (SynModuleDecl.Let (true, bindings, _)), Some checkInfo ->
+        match bindings with
+        | SynBinding (_, _, _, _, attributes, _, _, SynPat.LongIdent (SynLongIdent([ident], _, _), _, _, _, _, range), _, body, _, _, _) :: _ ->
+            let funcName = ident.idText
+        
+            let functionCallsItself =
+                checkInfo.GetAllUsesOfAllSymbolsInFile()
+                |> Seq.exists (fun usage -> 
+                    usage.Symbol.DisplayName = funcName 
+                    && ExpressionUtilities.rangeContainsOtherRange body.Range usage.Range)
+
+            if functionCallsItself then
+                let hasTailCallAttribute =
+                    attributes 
+                    |> List.collect (fun attrs -> attrs.Attributes) 
+                    |> List.exists 
+                        (fun attr -> 
+                            match attr.TypeName with
+                            | SynLongIdent([ident], _, _) ->
+                                ident.idText = "TailCall" || ident.idText = "TailCallAttribute"
+                            | _ -> false)
+                if hasTailCallAttribute then
+                    Array.empty
+                else
+                    { Range = range
+                      Message =
+                        String.Format(
+                            Resources.GetString "RulesEnsureTailCallDiagnosticsInRecursiveFunctions",
+                            funcName
+                        )
+                      SuggestedFix = None
+                      TypeChecks = list.Empty }
+                    |> Array.singleton
+            else
+                { Range = range
+                  Message =
+                    String.Format(
+                        Resources.GetString "RulesUnneededRecKeyword",
```

Comment:
@webwarrior-ws it's good that this rule doesn't copy a msg from another rule, but when I requested to respect DRY I actually meant it on a deeper level: this rule should call a function from the other rule, not repeat its logic here

Reply:
There is not much code to reuse and extracting common functionality to use in both rules will likely lead to more complexity.

Reply:
@webwarrior-ws then do this in a new commit and after I see the diff of it I will decide if we drop it or we keep it